src/Uno.UI/UI/Xaml/Documents/InlineCollection.skia.cs

```diff
@@ -437,13 +476,110 @@ internal void Draw(in DrawingSession session)
 						}
 					}
 
-					using var textBlob = _textBlobBuilder.Build();
-					canvas.DrawText(textBlob, 0, y + baselineOffsetY, paint);
+					{
+						if (RenderSelectionAndCaret && Selection is { } bg && bg.StartLine <= lineIndex && lineIndex <= bg.EndLine)
+						{
+							var spanStartingIndex = characterCountSoFar;
+
+							// x at this point is set to the right of the rightmost character ignoring spaces.
+
+							float left;
+							if (bg.StartIndex < spanStartingIndex)
+							{
+								// the selection starts from a previous span, so this span is selected from the very beginning
+								left = positions.Length > 0 ? positions[0].X : x;
+							}
+							else if (bg.StartIndex - spanStartingIndex < positions.Length)
+							{
+								// part or all of this span is selected
+								left = positions[bg.StartIndex - spanStartingIndex].X;
+							}
+							else
+							{
+								// this span is not a part of the selection, so we select nothing by making the left edge to the far right
+								left = x + justifySpaceOffset * segmentSpan.TrailingSpaces;
+							}
+
+							float right;
+							if (bg.EndIndex - spanStartingIndex < 0)
+							{
+								// this span is not a part of the selection, so we select nothing by making the left edge to the far left
+								right = positions.Length > 0 ? positions[0].X : x;
+							}
+							else if (bg.EndIndex - spanStartingIndex < positions.Length)
+							{
+								// part or all of this span is selected
+								right = positions[bg.EndIndex - spanStartingIndex].X;
+							}
+							else
+							{
+								// the selection ends after this span, so this span is selected to the very end
+								var allTrailingSpaces = segmentSpan.FullGlyphsLength - segmentSpan.GlyphsLength; // rendered and non-rendered trailing spaces
+								right = x + justifySpaceOffset * allTrailingSpaces;
+
+								if (bg.StartIndex != bg.EndIndex && SpanEndsInCR(segment, segmentSpan))
+								{
+									// fontInfo.SKFontSize / 3 is a heuristic width of a selected \r, which normally doesn't have a width
+									right += (segment.LineBreakAfter ? fontInfo.SKFontSize / 3 : 0);
+								}
+							}
+
+							if (Math.Abs(left - right) > 0.01 && fireEvents)
+							{
+								SelectionFound?.Invoke(new Rect(new Point(left, y - line.Height), new Point(right, y)));
+							}
+						}
+					}
+
+					if (glyphs.Length != 0)
+					{
+						using var textBlob = _textBlobBuilder.Build();
+						canvas.DrawText(textBlob, 0, y + baselineOffsetY, paint);
+					}
+
+					{
```

Comment:
Just to be sure, there's no other statement needed here. If not, let's remove the braces to avoid confusion.

Reply:
The braces are for scoping only. I found the variable names clear, so I wanted to reuse them below. I can remove them if necessary, but I find this clearer. Please confirm if so.

Reply:
Got it. If there starts to be conflicts this way, it may be a signal to split that method into multiple others.